man/unionfs.8
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.TH "unionfs" "8" "2025" "unionfs-fuse 3.7" ""
.SH "NAME"
unionfs\-fuse \- A userspace unionfs implementation
.SH "SYNOPSIS"
.B unionfs
\fB[\-o \fIoption1\fP \-o \fIoption2\fP ... \-o \fIoptionN\fP ]\fR
             \fBtop_branch:lower_branch:...:lowest_branch \fR
             \fBmount_point\fR
.SH "DESCRIPTION"
\fBunionfs\fR overlays several directories into one single mount point.
.PP
It first tries to access the file on the top branch and if the file does not exist
there, it continues on lower level branches.
If the user tries to modify a file on a lower level read\-only branch while
.B copy\-on\-write (cow)
mode is enabled, the file will be copied to a higher level read\-write branch.
.SH "OPTIONS"
Below is a summary of unionfs options
.TP
\fB\-o chroot=path
Path to chroot into. By using this option, unionfs
may be used for live CDs or live USB sticks, etc. So it can serve
"/" as filesystem. If you do not specify this option and try to use
it for "/", it will deadlock on calling 'pivot_root'.
If you set this option, you also need to specify the branches relatively
to the given chroot directory. See
.I examples/S01a-unionfs-live-cd.sh
for an example.
.TP
\fB\-o cow
Enable copy\-on\-write
.TP
\fB\-o hide_meta_files
In our unionfs root path we have a
.I .unionfs
directory that includes
metadata, such as hidden (deleted) files. This option makes this
directory invisible from readdir(), so for example
.B ls -la /union_root
will not show it. However, this directory is still there and
.B cd .unionfs
or
.B ls -l .unionfs
still work. Also, libfuse will create
.I \%.fuse_hidden*
files, if a file is open, but will be deleted. Those fuse meta files will
be invisible as well. This option is especially useful for package builders.
.TP
\fB\-d
Enable debugging for unionfs and libfuse. Useful for developers
if the code does not behave as expected. Debug information will be written
to stderr and a debug file (./unionfs_debug.log by default).
.TP
\fB\-o debug_file=file
Write unionfs debug information into that file.
.TP
\fB\-o max_files=number
Maximum number of open files. Most systems have a default limit of 1024
open files per process. For example if unionfs serves "/", applications
like KDE or GNOME might have many open files, making the unionfs process
reach this limit and unable to open further files. Suggested value for "/"
is >16000 or even >32000 files.
.TP
\fB\-o noinitgroups
Since version 0.23 without any effect, just left over for compatibility.
Might be removed in future versions.
.TP
\fB\-o relaxed_permissions
Usually we automatically add the libfuse option
.B \-o \%default_permissions
so that libfuse takes over permission checks. However, if running not
as root (so as UID\ !=\ 0 and GID\ !=\ 0), permissions on the underlying
filesystem are already sufficient. In order to prevent severe
security issues, this option is not allowed if running as root.
.TP
\fB\-o statfs_omit_ro
By default, blocks of all branches are counted in statfs() calls
(e.g. by 'df'). With this option, read-only branches will be omitted
from the summary of blocks. This may sound weird, but it actually fixes
"wrong" percentage of free space.
.TP
\fB\-o all_writable
Treat read\-only files and directories on backing filesystems as writable.
Implies copy\-on\-write.
.TP
.SH "Options to libfuse"
There are several further options available, which don't directly apply to
unionfs, but to libfuse. Please run
.B unionfs \-\-help
to see these. We already set
.B \-o \%default_permissions
option on our own.
.SH "EXAMPLES"
.Vb 5
\& unionfs \-o cow,max_files=32768 \e
\&              \-o allow_other,suid,dev,nonempty \e
\&              /u/host/etc=RW:/u/group/etc=RO:/u/common/etc=RO \e
\&              /u/union/etc
.Ve
.SH "Meta data"
Like other filesystems unionfs also needs to store meta data.
Well, presently only information about deleted files and directories need
to be stored, but in future releases more information might be required, e.g.
inode-numbers for persistent inode information.
Meta data information are saved and looked for in the .unionfs/
directories of each branch-root. So in the example above, these are
/u/host/etc/.unionfs, /u/group/etc/.unionfs and /u/common/etc/.unionfs.
Within these directories a complete directory structure may be found.
Example: If the admin decides to delete the file /etc/test/testfile, which
only exists in /u/unionfs/etc/test/testfile, unionfs can't delete this
file, since it is on a read-only branch. So instead the whiteout file
/u/host/etc/.unionfs/test/testfile_HIDDEN~ will be created. So on accessing
the union filesystem, test/testfile will not be visible.
Please also note that whiteout files/directories will only hide the files
in lower level branches. So for example whiteouts in the group directory
(/u/group/etc/.unionfs of the example above) will only hide file of the
common branch (/u/common/etc), but not these of the group and host branches.
Especially for diskless-booted environments it is rather useful for the admin
to create whiteout files him/her-self. For example one should blacklist
network re-initializations, /etc/mtab, /etc/nologin of the server and several
cron-scripts. This can be easily achieved by creating whiteout files for
these scripts in the group meta directory.
.SH "KNOWN ISSUES"
.Vb 5
\&1) Another issue is that presently there is no support for read-only branches
when copy-on-write is disabled, thus, -ocow is NOT specified! Support for
that might be added in later releases.
.Ve
.SH "AUTHORS"
.B unionfs\-fuse
Original implementation by Radek Podgorny <radek@podgorny.cz>
.SH "COPYRIGHT"
Radek Podgorny <radek\@podgorny.cz>, Bernd Schubert <bernd\-schubert\@gmx.de>
.SH "THANKS"
Many thanks to the author of the FUSE filesystem Miklos Szeredi.


NEWS
3.7
- fix chmod on symlink itself

3.6
- fix long standing bug with directory renames in cow mode (see #91)

3.5
- fix compilation on 32bit systems

3.4
- default to fuse2 api on macos when compiling

3.3
- fixed fortified compilation
- disabled logging to file since it had some unresolved locking issues
- add support for direct_io

3.2
- various nfs fixes
- various race condition fixes

3.1
- just fixes version mess (in manual file, cmake, etc.)

3.0
- libfuse3 support (and use is as new default)
- add branch-preserving logic
- make it compile under MacOS again
- minor fixes

2.2
- fixed timestamp precision lost on cow copy

2.1
- fixed Finder fails on MacOS
- Implement access() for "cow,relaxed_permissions" use case

2.0
- fixed NEWS file ;-)
- reverted meta directory name back to .unionfs (from .unionfs-fuse)
  this was never meant to be changed and slipped in in 1.0 release
- fixes and speed improvements

1.0
- add unionfsctl utility to control unionfs-fuse over ioctl
- better apple support
- dropped stats
- better test suite
- probably lots of other things, it's been a while since last release ;-)

0.26
- use BUILD_PATH instead of snprintf, which will
  return -ENAMETOOLONG if the path is too long instead of simply
  using a wrong path
- new options "-o relaxed_permissions" and "-o hide_meta_dir"
- included the debian directory
- Properly check on rmdir() if sub-branches are also empty
- Always compile with debug code, but only enable debugging on request
- New option -o debug_file
- Lots of bugs fixed in BUILD_PATH()
- Improved search of white-out files
- Properly fix Debian Bug#509516.
- More debug output.
- Add syslog support without the risk of possible dead locks
- Use fuse big-writes (and reads) if available

0.25
- Alternate way to specify branches
- Minor fixes

0.24
- Support for liveCDs / live USB sticks (-o chroot)
- Build-in support to change the maximum number of open files
- Added recursive directory COW.
- relative paths: Critical bug fix, had been completely broken in 0.23

0.23
- remove to_user() and to_root() calls for autorization and use fuse build-in
  default_permissions checks
- bugfix: rename created wrong whiteout type

0.22
- Fix a bug reported by Jens Hoelldampf <jens@hoelldampf.net>, in 0.21 cow
  didn't work for pathes.

0.21
- Fix a segmentation fault when COW was enabled, but no rw-branch was specified.
- Proper handling when ro-branches are on top of rw-branches.
- Disabled syslog entries for now, since it might cause deadlocks.
- Fixed a license issue with the elfhash, since the CPL is not compatible
  with the BSD license.
- Fix a bug when stat() was called instead of lstat().
- Many internal code changes (renaming of "root" to "branch").
- Better directory structure.
- Add this NEWS file.
- Add a man page.

0.20
- Fix a critical bug introduced in 0.19 when we disabled threadding.

0.19
- Several copy-on-write fixes (readdir, unlink, rmdir, rename).
- Whiteout files are now located in branch/.unionfs/ subdirectories.
- Disable threadding, since our current permissions-model is not compatible
  with the NPTL implementation and the Posix-Thread model in general.
- Add supplementary group support.

0.18
- First release with copy-on-write (COW) support.
- Many many internal code changes.


src/unionfs.c
/*
*
* This is offered under a BSD-style license. This means you can use the code for whatever you
* desire in any way you may want but you MUST NOT forget to give me appropriate credits when
* spreading your work which is based on mine. Something like "original implementation by Radek
* Podgorny" should be fine.
*
* License: BSD-style license
* Copyright: Radek Podgorny <radek@podgorny.cz>,
*            Bernd Schubert <bernd-schubert@gmx.de>
*/

#include <fuse.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/ioctl.h>

#include "unionfs.h"
#include "debug.h"
#include "opts.h"
#include "usyslog.h"

static struct fuse_opt unionfs_opts[] = {
	FUSE_OPT_KEY("chroot=%s,", KEY_CHROOT),
	FUSE_OPT_KEY("cow", KEY_COW),
	FUSE_OPT_KEY("preserve_branch", KEY_PRESERVE_BRANCH),
	FUSE_OPT_KEY("all_writable", KEY_ALL_WRITABLE),
	FUSE_OPT_KEY("debug_file=%s", KEY_DEBUG_FILE),
	FUSE_OPT_KEY("dirs=%s", KEY_DIRS),
	FUSE_OPT_KEY("--help", KEY_HELP),
	FUSE_OPT_KEY("-h", KEY_HELP),
	FUSE_OPT_KEY("hide_meta_dir", KEY_HIDE_METADIR),
	FUSE_OPT_KEY("hide_meta_files", KEY_HIDE_META_FILES),
	FUSE_OPT_KEY("max_files=%s", KEY_MAX_FILES),
	FUSE_OPT_KEY("noinitgroups", KEY_NOINITGROUPS),
	FUSE_OPT_KEY("relaxed_permissions", KEY_RELAXED_PERMISSIONS),
	FUSE_OPT_KEY("statfs_omit_ro", KEY_STATFS_OMIT_RO),
	FUSE_OPT_KEY("direct_io", KEY_DIRECT_IO),
	FUSE_OPT_KEY("--version", KEY_VERSION),
	FUSE_OPT_KEY("-V", KEY_VERSION),
	FUSE_OPT_END
};

int main(int argc, char *argv[]) {
	struct fuse_args args = FUSE_ARGS_INIT(argc, argv);

	// Initing syslog here is a problem because the log thread ends up exiting in non foreground mode (see https://github.com/rpodgorny/unionfs-fuse/issues/122)
	// init_syslog();
	uopt_init();

	if (fuse_opt_parse(&args, NULL, unionfs_opts, unionfs_opt_proc) == -1) RETURN(1);

	if (uopt.debug)	debug_init();

	if (!uopt.doexit) {
		if (uopt.nbranches == 0) {
			printf("You need to specify at least one branch!\n");
			RETURN(1);
		}
	}

	// enable fuse permission checks, we need to set this, even we we are
	// not root, since we don't have our own access() function
	uid_t uid = getuid();
	gid_t gid = getgid();
	bool default_permissions = true;

	if (uid != 0 && gid != 0 && uopt.relaxed_permissions) {
		default_permissions = false;
	} else if (uopt.relaxed_permissions) {
		// protect the user of a very critical security issue
		fprintf(stderr, "Relaxed permissions disallowed for root!\n");
		exit(1);
	}

	if (default_permissions) {
		if (fuse_opt_add_arg(&args, "-odefault_permissions")) {
			fprintf(stderr, "Severe failure, can't enable permssion checks, aborting!\n");
			exit(1);
		}
	}
	unionfs_post_opts();

#ifdef FUSE_CAP_BIG_WRITES
	/* libfuse > 0.8 supports large IO, also for reads, to increase performance
	 * We support any IO sizes, so lets enable that option */
	if (fuse_opt_add_arg(&args, "-obig_writes")) {
		fprintf(stderr, "Failed to enable big writes!\n");
		exit(1);
	}
#endif

	umask(0);
	int res = fuse_main(args.argc, args.argv, &unionfs_oper, NULL);
	RETURN(uopt.doexit ? uopt.retval : res);
}
